nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py

```diff
@@ -339,53 +475,11 @@ def inference_epoch_end(self, outputs):
         self.log('val_loss', averaged_loss, prog_bar=True, rank_zero_only=True)
         self.log('val_acc', val_acc, prog_bar=True, rank_zero_only=True)
 
-    def validation_step(self, batch, batch_idx, inference=False):
-        outcome = self.inference_step(batch, batch_idx, inference=inference)
-        return outcome
-
-    def validation_epoch_end(self, outputs):
-        self.inference_epoch_end(outputs)
-
-    def test_step(self, batch, batch_idx):
-        return self.validation_step(batch, batch_idx)
-
-    def test_epoch_end(self, outputs):
-        self.validation_epoch_end(outputs)
-
-    def build_virtual_prompt_dataset(
-        self, dataset_paths, batch_size, for_train, drop_last, shuffle, num_workers, pin_memory
-    ):
-        dataset = T5PromptLearningDataset(
-            datasets=dataset_paths,
-            tokenizer=self.tokenizer,
-            virtual_prompt_source=self.virtual_prompt_source,
-            task_templates=self.task_templates,
-            pseudo_tokens=self.pseudo_tokens,
-            pad_token_id=self.pad_token_id,
-            max_seq_length=self.cfg.data.get('max_seq_length', self.frozen_model.cfg.max_position_embeddings),
-            min_seq_length=self.cfg.data.get('min_seq_length', 1),
-            add_bos=self.cfg.data.get('add_bos', False),
-            add_eos=self.cfg.data.get('add_eos', True),
-            for_train=for_train,
-        )
-
-        rank = parallel_state.get_data_parallel_rank()
-        world_size = parallel_state.get_data_parallel_world_size()
-        sampler = torch.utils.data.distributed.DistributedSampler(
-            dataset, num_replicas=world_size, rank=rank, shuffle=shuffle
-        )
-
-        dataloader = torch.utils.data.DataLoader(
-            dataset,
-            collate_fn=dataset.collate_fn,
-            sampler=sampler,
-            batch_size=batch_size // world_size,
-            drop_last=drop_last,
-            num_workers=num_workers,
-            pin_memory=pin_memory,
-        )
-        print('build success', len(dataloader), dataset_paths)
-        return dataset, dataloader
+        # Save inference ready .nemo checkpoint version
+        if self.cfg.get("save_intermediate_nemo_file", True):
+            if self.lowest_val_loss is None or averaged_loss < self.lowest_val_loss:
+                self.save_checkpoint_as_nemo_file()
+                self.lowest_val_loss = averaged_loss
 
     def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: int = 0) -> Any:
```

Comment:
what's the difference between `inference_step` vs `predict_step` do we need both?

Reply:
`inference_step` was just used for validation. It used `embed_input_train` instead of `embed_input_inference`, it assumed a label was always present, it called `fwd_bwd_step` and calculated the validation loss. TBH, it should just be called `validation_step`. I renamed it to `validation_step` and did the same for `inference_epoch_end` -> `validation_epoch_end`.